### PR TITLE
Dem plane insertion

### PR DESCRIPTION
**Pull request by OGaboriault**
# Description of the problem
- The insertion in DEM can be a real nightmare when using a large number of particles. Particles can bounce back in the insertion zone which can cause "pop-corn' effect. Also, when the whole domain is close to be filled, big overlap between particle is unavoidable when inserting.
- This problem makes simulations slower or even can make them unusable went big overlap occurs. 

# Description of the solution
- Plane insertion method uses a group of cells, which are all in the same mathematical plane, to insert new particles. 
- On every insertion step, we check if those cells are individually empty. If a cell is empty, a new particle will be inserted at it center. This way, no overlap is accounted.
- Currently, the insertion stop when a certain number of particle is reach.    

# How Has This Been Tested?

- Insertion in a Hyper_cube with different refinement level have been used. 
- Different plane (point and normal vector) have be used in that Hyper_cube.  
- Test and application test have been created.

# Documentation
- insertion_info.rst --> Not done yet ( TODO ) 

# Future changes
- Documentation of this method need to be done.
- The stop criteria could the the filling of the domain and not the total number of particle inserted.

### CODE REVIEW

File: source/dem/plane_insertion.cc [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+template <int dim>
+void
+PlaneInsertion<dim>::find_inplane_cells(
+  const parallel::distributed::Triangulation<dim> &triangulation,
+  Point<3>                                         plane_point,
+  Tensor<1, 3>                                     plane_normal_vector)
+{ // Looping through cells
+  for (const auto &cell : triangulation.active_cell_iterators())
+    {
+      // If the cell is owned by owned by the processor
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:40]: owned by owned

File: source/dem/plane_insertion.cc [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      if (cell->is_locally_owned())
+        {
+          bool cell_in_plane = false;
+          // Initializing the values
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:44]: Wrong place?

File: source/dem/plane_insertion.cc [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+          // If the remained_particles_of_each_type become negative, this means
+          // that we are trying to insert to many particle in the simulation.
+          // Thus, we need to decrease the number of empty_cell at this time
+          // step. So...
+          if (remained_particles_of_each_type <= 0)
+            {
+              // ... we decrease the number of particle at the current
+              // processor, since it's the one who got the variable to become
+              // negative. Then...
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:162]: A bit vague, I would rephrase

File: source/dem/plane_insertion.cc [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                  number_of_particles_to_insert_per_core[j] = 0;
+                }
+              break;
+            }
+        }
+
+      // Now, every processor knows how many particle it needs to insert. They
+      // also know how many particle the other processor have to insert. This
+      // way, it's possible for a processor to find the correct range of
+      // particle id to use with the particle_counter variable.
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:183]:
```suggestion
      // Now, every processors know how many particles they need to insert. They
      // also know how many particles the other processors have to insert. This
      // way, it's possible for a processor to find the correct range of
      // particle id to use with the particle_counter variable.
```

File: source/dem/plane_insertion.cc [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                        particle_counter);
+
+      // We yet didn't say which empty cells won't be use anymore.
+      // empty_cells_on_proc needs to be shorter for some processor.
+      while (empty_cells_on_proc.size() <
+             number_of_particles_to_insert_per_core[this_mpi_process])
+        {
+          auto it = empty_cells_on_proc.end();
+          empty_cells_on_proc.erase(it);
+        }
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:199]: Here I think I would juste extract part of the vector like or remove in 1 line with something like:
empty_cells_on_proc.erase(empty_cells_on_proc.begin() + number_of_particles_to_insert_per_core[this_mpi_process], empty_cells_on_proc.end())

File: source/dem/plane_insertion.cc [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+      // Loop over the empty cells we kept.
+      for (const auto &cell : empty_cells_on_proc)
+        {
+          // Position of the center of the cell
+          Point<dim> insertion_location =
+            cells_centers.at(cell->active_cell_index());
+
+          // Generate the random point of insertion
+          insertion_location(0) +=
+            static_cast<double>(rand()) / static_cast<double>(RAND_MAX) *
```
Comment (acdaigneault) [inline source/dem/plane_insertion.cc:250]: Maybe you can have a variable =
            dem_parameters.insertion_info.random_number_range / static_cast<double>(RAND_MAX)
to multiply to each particle location random number since it's a constant

**Review by acdaigneault — COMMENTED** [review]
Good job Gabo! Very clean and easy. Some comments/suggestions here and there. I will test it when doc is done. Hopefully I will found a bug as a simili-revenge hehe

File: tests/dem/insertion_plane.cc [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  int refinement_number = 2;
+  tr.refine_global(refinement_number);
+
+  MappingQ<dim>            mapping(1);
+  DEMSolverParameters<dim> dem_parameters;
+
+  // Defining simulation general parameters
+  dem_parameters.insertion_info.insertion_plane_normal_vector =
+    Tensor<1, 3>({0., 1., 0.});
+  dem_parameters.insertion_info.insertion_plane_point = Point<3>({0., 2., 0});
```
[the remark this answers is not in the document]
Comment (OGaboriault) [reply tests/dem/insertion_plane.cc:60]: I can put " Point<3>({0., 1.75, 0});" , but that shouldn't change anything considering the way the code is implemented.

File: applications_tests/dem_3d/insert_plane_3d.output [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+22 0 0.01000 0.1625 -0.0060 -0.0737
+23 0 0.01000 0.4418 -0.0241 -0.3711
+24 0 0.01000 -0.0812 0.0159 0.4681
+25 0 0.01000 -0.3029 -0.0294 -0.0511
+26 0 0.01000 0.4390 -0.0166 0.4438
+27 0 0.01000 -0.3584 -0.0034 0.4630
+28 0 0.01000 0.2079 -0.0069 0.3979
+29 0 0.01000 -0.0357 -0.0020 -0.3063
+30 0 0.01000 0.2206 0.0303 0.1907
+31 0 0.01000 -0.0391 0.0187 0.2174
```
[the remark this answers is not in the document]
Comment (OGaboriault) [reply applications_tests/dem_3d/insert_plane_3d.output:43]: I think it has to do with the generation of the random number and the seed. 

Since both core have the same seed, I think that the n_th random number generated by core 1 will be the same as the n_th random number generated by core 2.

File: doc/source/parameters/dem/insertion_info.rst [hunk 9]
```diff
@@ -2,71 +2,76 @@
...
+    # If method = non_uniform or plane
     set insertion random number range                  = 0.75
     set insertion random number seed                   = 19
-  end
 
-* The ``insertion method`` parameter chooses the type of insertion. Acceptable choices are ``uniform`` and ``non_uniform``.
+    # If method = list
+    set list x                                         = 0.
+    set list y                                         = 0.
+    set list z                                         = 0.
```
Comment (blaisb) [inline doc/source/parameters/dem/insertion_info.rst:55]: Can you add the options for the velocity and angular velocity for the list insertion?

File: doc/source/parameters/dem/insertion_info.rst [hunk 10]
```diff
@@ -101,3 +104,24 @@ in a ``non_uniform`` insertion. :math:`{\epsilon}`, :math:`{\psi}`, and :math:`{
...
+
+* The ``insert plane point`` defines the point coordinates for the plane. Each component of this parameter represent the x, y and z directions, respectively.
+
+* The ``insertion plane normal vector`` define the normal vector component for the plane. of the  Each component of the parameter represent the x, y and z directions, respectively.
+
+--------------------
+List
+--------------------
+The List insertion method insert particles at precis coordinates. This method is preferred for small number of particles.
```
Comment (blaisb) [inline doc/source/parameters/dem/insertion_info.rst:119]:
```suggestion
The List insertion method insert particles at precise coordinates. This method is preferred for small number of particles.
```

**Review by blaisb — APPROVED** [review]
Chill change, good work. Only thing missing is boost signal?

File: doc/source/parameters/dem/insertion_info.rst (hunk 9 above)
Comment (OGaboriault) [reply doc/source/parameters/dem/insertion_info.rst:55]: I have look in the InsertionInfo::parse_parameters and in the ListInsertion<dim>::insert functions.

Velocity and angular velocity don't seems to be implemented for the list insertion method. I can you an other PR afterward. 
Maybe someone did implement this functionality in a branch, but it never got merge?

File: include/dem/plane_insertion.h [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Author: Olivier Gaboriault, Polytechnique Montreal, 2023
```
Comment (blaisb) [inline include/dem/plane_insertion.h:17]:
```suggestion
```
We have moved away from adding author names to files.

File: include/dem/plane_insertion.h [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+#ifndef plane_insertion_h
+#  define plane_insertion_h
+
+/**
+ * Insertion of particles using cells cut by a plane
+ *
+ * @note
+ *
+ * @author Olivier Gaboriault, Polytechnique Montreal 2023-
```
Comment (blaisb) [inline include/dem/plane_insertion.h:36]:
```suggestion
```

File: include/dem/plane_insertion.h [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+   * useful when dealing with a domain close to be fully filled with particle.
+   * In this situation, other insertion method have a high risk to create a big
+   * overlap between particles on insertion. The plane insertion method mitigate
+   * this risk by insertion at the center of empty cells.
+   *
+   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param triangulation Triangulation object used in the simulation.
+   */
+  PlaneInsertion<dim>(
+    const DEMSolverParameters<dim> &                 dem_parameters,
```
Comment (blaisb) [inline include/dem/plane_insertion.h:60]: Do you need to pass ALL DEM parameters or. you could just pass the plane parameter itself?

File: include/dem/plane_insertion.h [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+   * @param particle_handler The particle handler of particles which are being
+   * inserted.
+   * @param triangulation Triangulation to access the cells in which the
+   * particles are inserted.
+   * @param dem_parameters DEM parameters declared in the .prm file.
+   */
+  virtual void
+  insert(Particles::ParticleHandler<dim> &                particle_handler,
+         const parallel::distributed::Triangulation<dim> &triangulation,
+         const DEMSolverParameters<dim> &dem_parameters) override;
```
Comment (blaisb) [inline include/dem/plane_insertion.h:74]: Why do you need to pass the parameters everytime you insert? Isthat the way the interface is made? If so, we'll adress that in another subsequent PR

File: include/dem/plane_insertion.h [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    Tensor<1, 3>                                     plane_normal_vector);
+
+  /**
+   * @brief Store the location of the centers of all the cells that are in the plane
+   */
+  void
+  find_centers_of_inplane_cells();
+
+  void
+  Update_insertion_cell();
```
Comment (blaisb) [inline include/dem/plane_insertion.h:99]:
```suggestion
  update_insertion_cell();
```
we don't use capital letters in function names

File: source/dem/CMakeLists.txt [hunk 16]
```diff
@@ -91,5 +93,6 @@ add_library(lethe-dem
   ../../include/dem/visualization.h
   ../../include/dem/write_checkpoint.h)
 
+
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/dem/CMakeLists.txt:96]: Remains to be addressed

File: source/dem/plane_insertion.cc [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+  current_inserting_particle_type = 0;
+  particle_counter                = 0;
+
+  // Finding which cells are inplane
+  find_inplane_cells(
+    triangulation,
+    dem_parameters.insertion_info.insertion_plane_point,
+    dem_parameters.insertion_info.insertion_plane_normal_vector);
+
+  // // Finding the center of those cells
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/dem/plane_insertion.cc:25]: Not resolved yet???

File: source/dem/plane_insertion.cc [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+          // Loop over all the vertices of the cell
+          for (unsigned int vertex_id = 1; vertex_id < cell->n_vertices();
+               ++vertex_id)
+            {
+              // Vector that goes from the plane reference point to the vertex
+              // n-th
+              Tensor<1, 3> connecting_vector =
+                point_nd_to_3d(cell->vertex(vertex_id)) - plane_point;
+
+              // Normal distance between vertex n-th and the plan
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:66]:
```suggestion
              // Normal distance between vertex n-th and the plane
```

File: source/dem/plane_insertion.cc [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+                connecting_vector * plane_normal_vector;
+
+              // If the multiplication of the two distances is negative, then at
+              // least one of vertex of the cell is on both sides of the plane.
+              // If the multiplication is equal to zero, then one vertex is on
+              // the plane.
+              if (vertex_wall_distance_ref * vertex_wall_distance <= 0)
+                {
+                  cell_in_plane = true;
+                  break; // If so, we brake the loop, and go to the next cell
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:77]:
```suggestion
                  break; // If so, we break the loop, and go to the next cell
```

File: source/dem/plane_insertion.cc [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+PlaneInsertion<dim>::find_centers_of_inplane_cells()
+{
+  cells_centers.clear();
+  for (const auto &cell : plane_cells_for_insertion)
+    {
+      cells_centers.insert({cell->active_cell_index(), cell->center()});
+    }
+}
+
+// The main insertion function. insert_particle is utilized to insert the
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:99]:
```suggestion
// The main insertion function, insert_particle, is utilized to insert the
```

**Review by blaisb — CHANGES_REQUESTED** [review]
Some comments are still unaddressed, I have additional ones also

File: source/dem/plane_insertion.cc [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+          // we loop over the vector that we have created using the gather.
+          // Every integer in that vector gives the number of empty_cell of a
+          // certain processor.
+          for (unsigned int i = 0; i < total_mpi_process; ++i)
+            {
+              // We subtract those integer from the following variable
+              remained_particles_of_each_type -=
+                number_of_particles_to_insert_per_core[i];
+
+              // If the remained_particles_of_each_type variable become
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:172]:
```suggestion
              // If the remaining_particles_of_each_type variable become
```

File: source/dem/plane_insertion.cc [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+              // negative, this means that we are trying to insert to many
+              // particle in the simulation. Thus, we need to decrease the
+              // number of empty_cell we want to insert in at this time step.
+              if (remained_particles_of_each_type < 0)
+                {
+                  // At this point, we know the current processor have to many
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:178]:
```suggestion
                  // At this point, we know the current processor have too many
```

File: source/dem/plane_insertion.cc [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+            }
+          std::fill(remained_particles_for_every_proc.begin(),
+                    remained_particles_for_every_proc.end(),
+                    remained_particles_of_each_type);
+        }
+
+      // Now, processor zero knows how many particles will be inserted by every
+      // processor without exceeding the maximum number of particles.
+
+      // We now need to scatter 3 information processor, first how
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:223]:
```suggestion
      // We now need to scatter 3 information per processor, first how
```

File: source/dem/plane_insertion.cc [hunk 24]
```diff
@@ -0,0 +1,338 @@
...
+      // many particles it needs to insert. Second what is the ID of the
+      // first particle it will insert at this insertion step. The starting ID
+      // is mandatory, since we don't want two particles to have the same ID,
+      // and we don't want to use the global_insertion method because it uses an
+      // all_gather, which has a high computational cost. Third, the
+      // remained_particles_of_each_type variable need to be updated proc other
+      // than proc 0 will enter the insert function dans the simulation will
+      // stop.
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:231]:
```suggestion
      // all_gather, which has a high computational cost. Third, the
      // remained_particles_of_each_type variable needs to be updated on processors other
      // than processor 0 otherwise the insert function will become incoherent.
```

File: source/dem/plane_insertion.cc [hunk 25]
```diff
@@ -0,0 +1,338 @@
...
+      unsigned int starting_ID_on_proc =
+        Utilities::MPI::scatter(communicator, starting_IDs_on_every_proc, 0);
+
+      // This scatters remained_particles_of_each_type on every proc
+      remained_particles_of_each_type =
+        Utilities::MPI::scatter(communicator,
+                                remained_particles_for_every_proc,
+                                0);
+
+      // We yet didn't say which empty cells won't be use anymore.
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:249]:
```suggestion
      // We yet didn't choose which empty cells won't be used anymore.
```

File: source/dem/plane_insertion.cc [hunk 26]
```diff
@@ -0,0 +1,338 @@
...
+      while (empty_cells_on_proc.size() >
+             number_of_particles_to_insert_on_this_core)
+        {
+          auto it = empty_cells_on_proc
+                      .begin(); // Get the iterator of the first element
+          empty_cells_on_proc.erase(it); // Erase the first element
+        }
+
+      // There's probably a better ways to define the properties of a particle,
+      // but for now it works...
```
Comment (blaisb) [inline source/dem/plane_insertion.cc:263]: That's ok, you can remove this comment

File: tests/dem/insertion_non_uniform.cc [hunk 27]
```diff
@@ -80,7 +79,7 @@ test()
   Particles::ParticleHandler<dim> particle_handler(
     tr, mapping, DEM::get_number_properties());
 
-  // Calling uniform insertion
+  // Calling non uniform insertion
```
Comment (blaisb) [inline tests/dem/insertion_non_uniform.cc:82]:
```suggestion
  // Calling non-uniform insertion
```

File: tests/dem/insertion_plane.cc [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+* it, and/or modify it under the terms of the GNU Lesser General
+* Public License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+* The full text of the license can be found in the file LICENSE at
+* the top level of the Lethe distribution.
+*
+* ---------------------------------------------------------------------
+
+*
+ * Author: Olivier Gaboriault, Polytechnique Montreal, 2019-
```
Comment (blaisb) [inline tests/dem/insertion_plane.cc:17]:
```suggestion
```

File: tests/dem/insertion_plane.cc [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+                           */
+
+/**
+ * @brief Insertion particles using plane insertion class.
```
Comment (blaisb) [inline tests/dem/insertion_plane.cc:21]:
```suggestion
 * @brief Insertion of particles using the plane insertion class.
```

Requires a bit more of a description

**Review by blaisb — CHANGES_REQUESTED** [review]
Last batch of comments

File: include/dem/plane_insertion.h (hunk 13 above)
Comment (OGaboriault) [reply include/dem/plane_insertion.h:60]: Also need the random_number_range to initialize the variable Audrey talked about in one of her comment. 

But yes, we could replace ALL the parameters by only 3 (I think)

**Review by blaisb — APPROVED** [review]
